// webhook/controllers/sender.js
/*!
 * /controllers/sender.js 负责传送文件的中间件
 * Copyright(c) 2012
 * Author: jifeng <wade428@163.com>
 */

var urlLib = require('url');
var parse = urlLib.parse;
var fs = require('fs');
var path = require('path');
var http = require('http');



var Sender = function (options) {
  this.options = options;
};

Sender.prototype.middleware = function() {
  var self = this;
  var rundir = this.options.rundir;
  return function (req, res, next) {
    var urlObj = parse(req.url, true);
    var params = urlObj.query;
    
    var appname = params.appname;
    if (!appname) {
      return next(new Error('Invalid appname'));
    }
    var file = path.join(rundir, appname, 'out', appname);
    file = file + '.tgz';
    if (!fs.existsSync(file)) {
      return next(new Error(file + ' Not Exist'));
    }
    var dest = params.dest
    if (!dest) {
      return next(new Error('Invalid dest'));
    };

    var url = 'http://' + dest + '?appname=' + appname;
    var args = parse(url, true);
    var options = {
      hostname: args.hostname,
      port: args.port,
      path: args.path,
      method: 'POST',
      headers: {
        'Content-Type': 'application/octet-stream'
      }
    }

    var request = http.request(options, function(resProxy) {
      res.statusCode = resProxy.statusCode;
      return resProxy.pipe(res);
    });
    var stream = fs.createReadStream(file);
    stream.on('data', function (chunk) {
      request.write(chunk)
    });

    stream.on('end', function () {
      request.end();
    });
  };
};



module.exports = function (options) {
  return new Sender(options);
};
